src/Encoder.cpp

```diff
@@ -0,0 +1,193 @@
+        // header    | char[3]            | NBS packet header ☢ { 0xE2, 0x98, 0xA2 }
+        // length    | uint32_t           | Length of this packet after this value
+        // timestamp | uint64_t           | Timestamp the data was emitted in microseconds
+        // hash      | uint64_t           | the 64bit hash for the payload type
+        // payload   | char[length - 16]  | the data payload
+
+        // The size of our output timestamp, hash and data
+        uint32_t size = sizeof(packet.timestamp) + sizeof(packet.type) + packet.length;
+
+        uint64_t timestampMicros = packet.timestamp / 1000;
+
+#pragma pack(push, 1)
+        struct PacketHeader {
+            std::array<char, 3> header = {char(0xE2), char(0x98), char(0xA2)};
+            uint32_t size{0};
+            uint64_t timestamp{0};
+            uint64_t hash{0};
+
+            PacketHeader(const uint32_t& size, const uint64_t& timestamp, const uint64_t& hash)
+                : size(size), timestamp(timestamp), hash(hash){};
+        };
+#pragma pack(pop)
+
+        // Build the header section of the packet
+        std::vector<uint8_t> packetBytes(sizeof(PacketHeader), '\0');
+        new (packetBytes.data()) PacketHeader(size, timestampMicros, packet.type);
+
+        // Append the payload bytes to the packet
+        const auto* data = reinterpret_cast<const uint8_t*>(packet.payload);
+        packetBytes.insert(packetBytes.end(), data, data + packet.length);
+
+        // Write out the packet to the nbs file
+        outputFile.get()->write(reinterpret_cast<const char*>(packetBytes.data()), int64_t(packetBytes.size()));
+
+        return packetBytes.size();
+    }
+
+    void Encoder::WriteIndex(const Packet& packet, const uint32_t& size) {
+
+        // NBS Index File Format
+        // Name      | Type               |  Description
+        // ------------------------------------------------------------
+        // hash      | uint64_t           | the 64bit hash for the payload type
+        // subtype   | uint32_t           | the id field of the payload
+        // timestamp | uint64_t           | Timestamp of the message or the emit timestamp in nanoseconds
+        // offset    | uint64_t           | offset to start of radiation symbol ☢
+        // size      | uint32_t           | Size of the whole packet from the radiation symbol
+
+#pragma pack(push, 1)
+        struct PacketIndex {
+            uint64_t hash{0};
+            uint32_t subtype{0};
+            uint64_t timestamp{0};
+            uint64_t offset{0};
+            uint32_t size{0};
+
+            PacketIndex(const uint64_t& hash,
+                        const uint32_t& subtype,
+                        const uint64_t& timestamp,
+                        const uint64_t& offset,
+                        const uint32_t& size)
+                : hash(hash), subtype(subtype), timestamp(timestamp), offset(offset), size(size){};
+        };
+#pragma pack(pop)
```

Comment:
I think this form of packing is a better option than pragma
```suggestion
        struct PacketIndex {
            uint64_t hash{0};
            uint32_t subtype{0};
            uint64_t timestamp{0};
            uint64_t offset{0};
            uint32_t size{0};

            PacketIndex(const uint64_t& hash,
                        const uint32_t& subtype,
                        const uint64_t& timestamp,
                        const uint64_t& offset,
                        const uint32_t& size)
                : hash(hash), subtype(subtype), timestamp(timestamp), offset(offset), size(size){};
        } __attribute__((packed));
```